designs/nodepool-requirement-flexibility.md

```diff
@@ -0,0 +1,185 @@
+Karpenter’s scheduling algorithm uses a well-known bin-packing algorithm ([First Fit Descending](https://en.wikipedia.org/wiki/First-fit-decreasing_bin_packing)) to pack as many pods onto a set of instance type options as possible. This algorithm will necessarily continue packing pods onto Preflight Nodes so long as you have at least one instance type option that can fulfill all the pod requests.
+
+This methodology works fine for on-demand instance types, but starts to break down when you are requesting spot capacity. If you are launching spot capacity and want to ensure that you will not launch an instance that will immediately get interrupted after launch, you need to make sure that you have enough launch options in your launch request that your likelihood for launching a low-availability instance will be slim. Karpenter currently has no mechanism to ensure that today.
+
+This RFC proposes an additional key `minValues` in the NodePool requirements block, allowing Karpenter to be aware of user-specified flexibility minimums while scheduling pods to a cluster. If Karpenter cannot meet this minimum flexibility when scheduling a pod, it will fail the scheduling loop for that NodePool, either falling back to another NodePool which meets the pod requirements or failing scheduling the pod altogether.
+
+```
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+spec:
+  template:
+    spec:
+      requirements:
+        - key: karpenter.kwok.sh.instance-family
+          operator: In
+          values: ["c", "m", "r"]
+          minValues: 2
+        - key: node.kubernetes.io/instance-type
+          operator: Exists
+          minValues: 10
+```
+
+## Use-Cases
+
+1. I want to configure Karpenter to enforce more flexibility when launching spot nodes, ensuring that I am constantly using enough instance types that I get a high-availability instance type
+2. I want to increase the number of nodes that Karpenter creates by ensuring that Karpenter doesn’t over-pack pods onto nodes when it’s scheduling
+3. I want to ensure that application pods scheduling to this NodePool *must* be flexible to different instance types, architectures, etc.
+
+## Solutions
+
+### [Recommended] Solution 1: `minValues` field in the `spec.requirements`  section of NodeClaim Template (becomes part of the template on the NodePool)
+
+```
+apiVersion: karpenter.sh/v1beta1
+kind: NodePool
+spec:
+  template:
+    spec:
+      requirements:
+      - key: karpenter.kwok.sh.instance-family
+        operator: In
+        values: ["c", "m", "r"]
+        minValues: 2
+      - key: node.kubernetes.io/instance-type
+        operator: Exists
+        minValues: 10
+```
+
+We would add a new `minValues` field to the requirements section of the NodeClaim template. By adding this section to the NodeClaim template, we inform the scheduler that it should stop continuing to bin-pack pods onto Preflight NodeClaims and should, instead, create another Preflight NodeClaim that will schedule that pod.
```

Comment:
If I have zero nodes and try to launch a new node, does that new node satisfy `minValues: 2`? If not: does Karpenter avoid using the NodePool at all (maybe there needs to be a fallback NodePool to handle a mass node failure?)

&mdash; or: does Karpenter launch a second node to satisfy the NodePool `.spec`? If not, why not? “Minimum values” implies this.

(I think the proposed field name is not quite right)

Reply:
> “Minimum values” implies this.

I'm not I quite follow. This field name is talking about the minimum values that are required for a request of a NodeClaim, not the number of values that need to be satisfied across nodes within a NodePool. For that, it seems like we need a concept of `topologySpreadConstraints` on the NodePool to tell Karpenter, "give me something from this set of instance types, make sure that my request has at least two instance types in it, and I also want to make sure that you spread me across instance types with a minDomains of 3"